libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select.ts

```diff
@@ -0,0 +1,344 @@
+   * Can be configured globally through the `DT_MULTI_SELECT_DEFAULT_OPTIONS` token.
+   */
+  @Input()
+  get autoActiveFirstOption(): boolean {
+    return this._autoActiveFirstOption;
+  }
+  set autoActiveFirstOption(value: boolean) {
+    this._autoActiveFirstOption = coerceBooleanProperty(value);
+  }
+  private _autoActiveFirstOption: boolean;
+
+  /**
+   * Specify the width of the multi-select panel.  Can be any CSS sizing value, otherwise it will
+   * match the width of its host.
+   */
+  @Input() panelWidth: string | number;
+
+  /** Function that maps an option's control value to its display value in the trigger. */
+  @Input() displayWith: ((value: T) => string) | null = null;
+
+  /** Options or groups to be displayed */
+  @Input()
+  get optionsOrGroups(): Array<T & DtNodeDef> {
+    return this._optionsOrGroups ?? [];
+  }
+  set optionsOrGroups(opts: Array<T & DtNodeDef>) {
+    this._optionsOrGroups = !!opts ? opts : [];
+    this._filterOptions();
+  }
+  _optionsOrGroups: Array<T & DtNodeDef> = [];
+  _filteredOptionsOrGroups: Array<T & DtNodeDef> = [];
+
+  /** Value input by the user used to highlight and filter */
+  @Input()
+  get inputValue(): string {
+    return this._inputValue || '';
+  }
+  set inputValue(value: string) {
+    this._inputValue = value.toLowerCase();
+    this._checkApplyDisable();
+    this._filterOptions();
+  }
+  _inputValue: string;
+
+  /** Event that is emitted when the filter-field multiSelect panel is opened. */
+  @Output() readonly opened = new EventEmitter<void>();
+
+  /** Event that is emitted when the filter-field multiSelect panel is closed. */
+  @Output() readonly closed = new EventEmitter<void>();
+
+  /** Event that is emitted whenever an option from the list is selected. */
+  @Output() readonly multiSelectSubmitted = new EventEmitter<
+    DtFilterFieldMultiSelectSubmittedEvent<T>
+  >();
+
+  /** Unique ID to be used by filter-field multiSelect trigger's "aria-owns" property. */
+  id = `dt-filter-field-multiSelect-${_uniqueIdCounter++}`;
+
+  /** Whether the filter-field multiSelect panel is open. */
+  get isOpen(): boolean {
+    return this._isOpen;
+  }
+  /** @internal Whether the filter-field multiSelect panel is open. */
+  _isOpen = false;
+
+  /** @internal */
+  @ViewChild(TemplateRef, { static: true }) _template: TemplateRef<{}>;
+
+  /**
+   * @internal Reference to the panel which will be created in the overlay.
+   */
+  @ViewChild('panel') _panel: ElementRef;
+
+  /** @internal Querylist of options to be used in keymanager */
+  @ViewChildren(DtOption)
+  _options = new QueryList<DtOption<T>>();
+
+  /**
+   * @internal
+   * Manages active item in option list based on key events.
+   */
+  _keyManager: ActiveDescendantKeyManager<DtOption<T>>;
+
+  /** @internal */
+  _portal: TemplatePortal;
+
+  /** @internal Holds the current values of the input field for the from value */
+  _currentSelection: T[] = [];
```

Comment:
Angular CDK can help you with selections. It provides a `SelectionModel` class that you can use. Especially powerful for multi selects as you do.
https://github.com/angular/components/blob/master/src/cdk/collections/selection-model.ts